File: macos/Sources/MalachiCore/Wizard/CertTrust.swift
// SPDX-FileCopyrightText: 2026 Vladislav Janeček
// SPDX-License-Identifier: GPL-3.0-or-later

// ui/internal/certtrust: the certificate of a failed TLS connection,
// cleaned for display, and the rules for trusting it by pinning its
// SHA-256 fingerprint to an IMAP/SMTP endpoint (docs/security.md §7). No
// texts: the sentences are the wizard's and the window's.

import Foundation

/// The certtrust package: a namespace, so the Go names map 1:1
/// (`certtrust.Details` → `CertTrust.details`).
public enum CertTrust {
    /// certtrust.maxText: the longest certificate string shown, in UTF-8
    /// bytes (the daemon's own cap).
    public static let maxText = 128

    /// certtrust.maxList: the most names of a certificate shown.
    public static let maxList = 8

    /// certtrust.Problem: the cleaned details of a `tlsError`.
    public struct Problem: Sendable, Equatable {
        /// A reason this client does not know is `other`.
        public var reason: TLSErrorReason
        /// The certificate the server presented; nil when there was none
        /// or its fingerprint was not 64 hex digits.
        public var cert: CertificateInfo?
        /// pinMismatch: the pinned fingerprint (64 lowercase hex); "" otherwise.
        public var expected: String

        public init(reason: TLSErrorReason, cert: CertificateInfo? = nil, expected: String = "") {
            self.reason = reason
            self.cert = cert
            self.expected = expected
        }

        /// certtrust.Problem.Category.
        public var category: Category { CertTrust.categoryOf(reason) }
    }

    /// certtrust.Category: what kind of TLS failure a reason is.
    public enum Category: Sendable, Equatable {
        /// The connection could not be secured before a certificate was
        /// judged: handshake, starttlsUnavailable, tlsRequired.
        case connection
        /// The certificate was refused: untrusted, hostnameMismatch,
        /// expired, notYetValid, invalid, other, and any unknown reason.
        case certificate
        /// Not the pinned certificate: pinMismatch.
        case changed
    }

    /// certtrust.CategoryOf: sorts a reason; unknown reasons count as
    /// `other`.
    public static func categoryOf(_ reason: TLSErrorReason) -> Category {
        switch normalizeReason(reason) {
        case .handshake, .starttlsUnavailable, .tlsRequired:
            return .connection
        case .pinMismatch:
            return .changed
        default:
            return .certificate
        }
    }

    /// certtrust.NormalizeReason: `r` when it is a reason of docs/api.md §2,
    /// `other` for anything else, the empty reason included.
    public static func normalizeReason(_ r: TLSErrorReason) -> TLSErrorReason {
        TLSErrorReason.all.contains(r) ? r : .other
    }

    /// certtrust.Details: the TLS details of `e` with every string cleaned
    /// again for display (the daemon's cleaning is not relied on): control
    /// and format characters removed, trimmed, at most `maxText` bytes,
    /// lists of at most `maxList` non-empty entries. A certificate whose
    /// fingerprint is not 64 hex digits is dropped whole; so is such an
    /// expected fingerprint. nil when `e` is not a `tlsError`, has no
    /// details or no reason.
    public static func details(_ e: RPCError?) -> Problem? {
        guard let raw = tlsErrorData(e), !raw.reason.rawValue.isEmpty else { return nil }
        var p = Problem(reason: normalizeReason(raw.reason), cert: cleanCertificate(raw.certificate))
        p.expected = normalizeCertificateSHA256(raw.expectedSha256 ?? "") ?? ""
        return p
    }

    /// certtrust.cleanCertificate: nil without a valid fingerprint.
    private static func cleanCertificate(_ c: CertificateInfo?) -> CertificateInfo? {
        guard let c, let sum = normalizeCertificateSHA256(c.sha256) else { return nil }
        return CertificateInfo(
            sha256: sum,
            subject: nonEmpty(cleanText(c.subject ?? "")),
            issuer: nonEmpty(cleanText(c.issuer ?? "")),
            dnsNames: cleanList(c.dnsNames),
            ipAddresses: cleanList(c.ipAddresses),
            notBefore: c.notBefore,
            notAfter: c.notAfter,
            selfSigned: c.selfSigned
        )
    }

    /// certtrust.cleanText: `s` without the replacement character (what
    /// invalid UTF-8 became), control characters (Cc, newlines included),
    /// format characters (Cf, such as U+202E, which reverses what follows)
    /// and the line and paragraph separators (Zl, Zp), trimmed of
    /// White_Space (strings.TrimSpace) and cut to `maxText` bytes on a
    /// character boundary, then trimmed again.
    public static func cleanText(_ s: String) -> String {
        var kept = String.UnicodeScalarView()
        for c in s.unicodeScalars {
            switch c.properties.generalCategory {
            case .control, .format, .lineSeparator, .paragraphSeparator:
                continue
            default:
                if c.value != 0xFFFD {
                    kept.append(c)
                }
            }
        }
        let out = trimSpace(String(kept))
        if out.utf8.count <= maxText {
            return out
        }
        var cut = String.UnicodeScalarView()
        var bytes = 0
        for c in out.unicodeScalars {
            let n = String(c).utf8.count
            if bytes + n > maxText {
                break
            }
            bytes += n
            cut.append(c)
        }
        return trimSpace(String(cut))
    }

    /// certtrust.cleanList: the cleaned, non-empty entries, at most
    /// `maxList` of them.
    public static func cleanList(_ list: [String]) -> [String] {
        var out: [String] = []
        for s in list {
            let c = cleanText(s)
            if c.isEmpty {
                continue
            }
            out.append(c)
            if out.count == maxList {
                break
            }
        }
        return out
    }

    /// certtrust.FormatFingerprint: a SHA-256 fingerprint as upper-case hex
    /// in 16 groups of four joined by spaces ("AB12 CD34 …"), as it is
    /// compared by eye; "" for anything that is not a fingerprint.
    public static func formatFingerprint(_ hex: String) -> String {
        guard let norm = normalizeCertificateSHA256(hex) else { return "" }
        var out = ""
        for (i, c) in norm.uppercased().enumerated() {
            if i > 0, i % 4 == 0 {
                out.append(" ")
            }
            out.append(c)
        }
        return out
    }

    /// certtrust.Pinnable: the endpoint may carry a pin at all: TLS or
    /// STARTTLS with a password (never plaintext, and the tokens of an
    /// oauth2 endpoint never go to a pinned server).
    public static func pinnable(_ sc: ServerConfig) -> Bool {
        (sc.security == .tls || sc.security == .starttls) && sc.authMethod == .password
    }

    /// certtrust.Trustable: the problem of a failed endpoint of
    /// `account.test` whose certificate the user may trust: a certificate
    /// was presented and judged (not a connection failure), the endpoint
    /// may carry a pin, and it does not pin this very certificate already.
    /// A pinMismatch offers the new certificate. nil otherwise (and for
    /// Graph, which the wizard never asks about).
    public static func trustable(_ res: EndpointTestResult?, _ sc: ServerConfig) -> Problem? {
        guard let res, !res.ok, let p = details(res.error), let cert = p.cert, p.category != .connection,
              pinnable(sc) else { return nil }
        if cert.sha256 == normalizeCertificateSHA256(sc.certificateSha256 ?? "") {
            return nil
        }
        return p
    }

    /// certtrust.SameCertificate: both endpoints presented one and the same
    /// certificate (a mail bridge serving IMAP and SMTP): one confirmation
    /// pins both.
    public static func sameCertificate(_ a: CertificateInfo?, _ b: CertificateInfo?) -> Bool {
        guard let a, let b, let x = normalizeCertificateSHA256(a.sha256), let y = normalizeCertificateSHA256(b.sha256) else {
            return false
        }
        return x == y
    }

    /// certtrust.FromSyncState: the certificate problem of an account: it is
    /// offline or failed and its last error is a `tlsError` about the
    /// certificate (with or without the certificate itself). nil for
    /// anything else, a connection failure included (that stays
    /// "Offline").
    public static func fromSyncState(_ s: SyncState) -> Problem? {
        guard s.status == .offline || s.status == .error, let p = details(s.error), p.category != .connection else {
            return nil
        }
        return p
    }

    /// certtrust.KeepPin: `old`'s pin for `new` only while the host
    /// (trimmed, ignoring case, as DNS does) and the port are those it was
    /// trusted for and `new` may carry a pin; "" otherwise. The wizard keeps
    /// the configuration a pin was set for, so a changed host drops the pin
    /// and typing the old one again restores it; a pin never moves to
    /// another server.
    public static func keepPin(_ old: ServerConfig, _ cur: ServerConfig) -> String {
        guard let pin = normalizeCertificateSHA256(old.certificateSha256 ?? ""), pinnable(cur), old.port == cur.port,
              foldKey(trimSpace(old.host)) == foldKey(trimSpace(cur.host)) else { return "" }
        return pin
    }

    /// The case-folded scalars of `s`, compared literally as Go compares
    /// strings (strings.EqualFold, strings.ToLower keys): no canonical
    /// equivalence, so a composed and a decomposed name differ there and
    /// here, which Swift's `==` on String would not.
    static func foldKey(_ s: String) -> [Unicode.Scalar] {
        Array(s.lowercased().unicodeScalars)
    }

    /// certtrust.ServerName: "host:port" for the texts, an IPv6 literal in
    /// brackets (net.JoinHostPort).
    public static func serverName(_ sc: ServerConfig) -> String {
        let h = trimSpace(sc.host)
        if h.contains(":") {
            return "[" + h + "]:" + String(sc.port)
        }
        return h + ":" + String(sc.port)
    }

    /// certtrust.IssuedTo: whom the certificate was issued to: the subject
    /// on the first line, the DNS names and then the IP addresses on the
    /// second (joined with ", ", duplicates and the subject itself left out,
    /// ignoring case); "" when there is nothing.
    public static func issuedTo(_ c: CertificateInfo) -> String {
        var lines: [String] = []
        let subject = cleanText(c.subject ?? "")
        if !subject.isEmpty {
            lines.append(subject)
        }
        var seen: Set<[Unicode.Scalar]> = [foldKey(subject)]
        var names: [String] = []
        for n in cleanList(c.dnsNames) + cleanList(c.ipAddresses) {
            let key = foldKey(n)
            if seen.contains(key) {
                continue
            }
            seen.insert(key)
            names.append(n)
        }
        if !names.isEmpty {
            lines.append(names.joined(separator: ", "))
        }
        return lines.joined(separator: "\n")
    }

    /// Go's "" as nil.
    private static func nonEmpty(_ s: String) -> String? {
        s.isEmpty ? nil : s
    }

    /// strings.TrimSpace: only White_Space goes.
    private static func trimSpace(_ s: String) -> String {
        let scalars = s.unicodeScalars
        guard let first = scalars.firstIndex(where: { !$0.properties.isWhitespace }),
              let last = scalars.lastIndex(where: { !$0.properties.isWhitespace }) else { return "" }
        return String(scalars[first...last])
    }
}

File: ui/internal/certtrust/certtrust.go
// SPDX-FileCopyrightText: 2026 Vladislav Janeček
// SPDX-License-Identifier: GPL-3.0-or-later

// Package certtrust decides what the UI shows and offers when an IMAP or
// SMTP server's certificate is refused: the details of a tlsError
// (docs/api.md §2, TLSErrorData), whether the account assistant may offer
// to trust the certificate by pinning its SHA-256 fingerprint
// (ServerConfig.certificateSha256), when a pin survives an edit of the
// server settings, and which account states are certificate problems. It
// holds no GTK and no user-facing text, so its rules are tested without a
// display; the assistant, the sidebar, the banner and the Accounts page act
// on its answers, and the macOS client mirrors them one to one.
package certtrust

import (
	"net"
	"strconv"
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/schotek/malachi/backend/pkg/api"
)

// Caps of CertificateInfo (docs/api.md §2). The backend enforces them; the
// UI enforces them again because the text is the server's.
const (
	maxTextBytes = 128
	maxListItems = 8
)

// Problem is a cleaned tlsError: why the connection failed, the server's
// certificate when it presented a usable one (nil otherwise), and for
// api.TLSPinMismatch the pinned fingerprint ("" when absent). Reason is one
// of the api.TLS* reasons; an unknown one is api.TLSOther. Every string of
// Cert is plain text without control or format characters, SHA256 and
// Expected are 64 lowercase hex digits.
type Problem struct {
	Reason   api.TLSErrorReason
	Cert     *api.CertificateInfo
	Expected string
}

// Category groups the reasons by what the user can do about them.
type Category int

const (
	// Connection: no certificate was judged (handshake failure, STARTTLS
	// not offered, the server demands TLS before login). Nothing to trust;
	// the account shows as offline.
	Connection Category = iota
	// Certificate: the server's certificate was refused (untrusted issuer,
	// another name, expired, not yet valid, invalid, anything else the
	// system verifier refused, and any reason this client does not know).
	Certificate
	// Changed: the endpoint pins a certificate and the server presented
	// another one.
	Changed
)

// CategoryOf sorts a reason; unknown reasons count as api.TLSOther.
func CategoryOf(r api.TLSErrorReason) Category {
	switch NormalizeReason(r) {
	case api.TLSHandshake, api.TLSStartTLSUnavail, api.TLSRequired:
		return Connection
	case api.TLSPinMismatch:
		return Changed
	}
	return Certificate
}

// Category is the category of p's reason.
func (p Problem) Category() Category { return CategoryOf(p.Reason) }

// NormalizeReason returns r when it is a reason of docs/api.md §2 and
// api.TLSOther for anything else, the empty reason included.
func NormalizeReason(r api.TLSErrorReason) api.TLSErrorReason {
	switch r {
	case api.TLSUntrusted, api.TLSHostnameMismatch, api.TLSExpired, api.TLSNotYetValid,
		api.TLSInvalid, api.TLSOther, api.TLSHandshake, api.TLSStartTLSUnavail,
		api.TLSRequired, api.TLSPinMismatch:
		return r
	}
	return api.TLSOther
}

// Details reads the TLS details of e. false when e is not a tlsError or
// carries no details (no data, or data without a reason). The reason is
// normalised, every certificate string cleaned and capped, lists capped;
// a certificate whose fingerprint is not 64 hex digits is dropped whole,
// an invalid expected fingerprint becomes "".
func Details(e *api.Error) (Problem, bool) {
	d, ok := api.TLSErrorDataOf(e)
	if !ok || d.Reason == "" {
		return Problem{}, false
	}
	p := Problem{Reason: NormalizeReason(d.Reason), Cert: cleanCertificate(d.Certificate)}
	if exp, ok := api.NormalizeCertificateSHA256(d.ExpectedSHA256); ok {
		p.Expected = exp
	}
	return p, true
}

func cleanCertificate(c *api.CertificateInfo) *api.CertificateInfo {
	if c == nil {
		return nil
	}
	sum, ok := api.NormalizeCertificateSHA256(c.SHA256)
	if !ok {
		return nil
	}
	return &api.CertificateInfo{
		SHA256:      sum,
		Subject:     cleanText(c.Subject),
		Issuer:      cleanText(c.Issuer),
		DNSNames:    cleanList(c.DNSNames),
		IPAddresses: cleanList(c.IPAddresses),
		NotBefore:   c.NotBefore,
		NotAfter:    c.NotAfter,
		SelfSigned:  c.SelfSigned,
	}
}

// cleanText drops invalid UTF-8, control characters (Cc, newlines
// included), format characters (Cf, such as U+202E, which reverses what
// follows) and the line and paragraph separators (Zl U+2028, Zp U+2029),
// trims space and caps the result at maxTextBytes on a rune boundary,
// trimming again after the cut.
func cleanText(s string) string {
	var b strings.Builder
	for _, r := range s {
		if r == utf8.RuneError || unicode.IsControl(r) || unicode.In(r, unicode.Cf, unicode.Zl, unicode.Zp) {
			continue
		}
		b.WriteRune(r)
	}
	out := strings.TrimSpace(b.String())
	if len(out) <= maxTextBytes {
		return out
	}
	cut := maxTextBytes
	for cut > 0 && !utf8.RuneStart(out[cut]) {
		cut--
	}
	return strings.TrimSpace(out[:cut])
}

// cleanList cleans every entry, drops the empty ones and keeps at most
// maxListItems.
func cleanList(list []string) []string {
	var out []string
	for _, s := range list {
		if len(out) == maxListItems {
			break
		}
		if s = cleanText(s); s != "" {
			out = append(out, s)
		}
	}
	return out
}

// FormatFingerprint shows a SHA-256 fingerprint as 16 groups of 4
// upper-case hex digits separated by single spaces ("AB12 CD34 …"); "" when
// hex is not a fingerprint NormalizeCertificateSHA256 accepts.
func FormatFingerprint(hex string) string {
	sum, ok := api.NormalizeCertificateSHA256(hex)
	if !ok {
		return ""
	}
	sum = strings.ToUpper(sum)
	var b strings.Builder
	for i := 0; i < len(sum); i += 4 {
		if i > 0 {
			b.WriteByte(' ')
		}
		b.WriteString(sum[i : i+4])
	}
	return b.String()
}

// Pinnable reports an endpoint configuration that may pin a certificate:
// TLS or STARTTLS, signing in with a password (docs/api.md §4.1; tokens of
// an oauth2 endpoint never go to a pinned server).
func Pinnable(sc api.ServerConfig) bool {
	return (sc.Security == api.SecurityTLS || sc.Security == api.SecuritySTARTTLS) &&
		sc.AuthMethod == api.AuthPassword
}

// Trustable decides whether the assistant offers to trust the certificate
// of an endpoint that failed the connection test with configuration sc:
// the test failed with a tlsError whose details name a certificate with a
// valid fingerprint, the reason is about that certificate (Certificate or
// Changed, never Connection), sc is Pinnable, and the certificate is not
// the one sc already pins. The Problem is what the confirmation shows.
func Trustable(res *api.EndpointTestResult, sc api.ServerConfig) (Problem, bool) {
	if res == nil || res.OK || res.Error == nil {
		return Problem{}, false
	}
	p, ok := Details(res.Error)
	if !ok || p.Cert == nil || p.Category() == Connection || !Pinnable(sc) {
		return Problem{}, false
	}
	if pinned, ok := api.NormalizeCertificateSHA256(sc.CertificateSHA256); ok && pinned == p.Cert.SHA256 {
		return Problem{}, false
	}
	return p, true
}

// SameCertificate reports two certificates with the same valid fingerprint
// (IMAP and SMTP of a mail bridge presenting one certificate: one
// confirmation trusts both).
func SameCertificate(a, b *api.CertificateInfo) bool {
	if a == nil || b == nil {
		return false
	}
	x, ok := api.NormalizeCertificateSHA256(a.SHA256)
	if !ok {
		return false
	}
	y, ok := api.NormalizeCertificateSHA256(b.SHA256)
	return ok && x == y
}

// FromSyncState reports an account whose last connection attempt failed on
// its server's certificate: status offline or error with a tlsError whose
// details are not of Category Connection. The certificate itself is not
// required (the sidebar and the banner only need the category).
func FromSyncState(s api.SyncState) (Problem, bool) {
	if s.Status != api.SyncOffline && s.Status != api.SyncError {
		return Problem{}, false
	}
	p, ok := Details(s.Error)
	if !ok || p.Category() == Connection {
		return Problem{}, false
	}
	return p, true
}

// KeepPin is the pin an endpoint keeps after its settings changed from old
// to cur: old's pin (normalised) while the host (trimmed, case-insensitive)
// and the port are the same and cur is still Pinnable; "" otherwise, and
// when old has no valid pin. A pin is trusted for one server, never for
// whatever the endpoint is changed to.
func KeepPin(old, cur api.ServerConfig) string {
	pin, ok := api.NormalizeCertificateSHA256(old.CertificateSHA256)
	if !ok || !Pinnable(cur) || old.Port != cur.Port ||
		!strings.EqualFold(strings.TrimSpace(old.Host), strings.TrimSpace(cur.Host)) {
		return ""
	}
	return pin
}

// ServerName is the endpoint as the confirmation names it: "host:port",
// with an IPv6 literal in brackets.
func ServerName(sc api.ServerConfig) string {
	return net.JoinHostPort(strings.TrimSpace(sc.Host), strconv.Itoa(sc.Port))
}

// IssuedTo is the "Issued to" value of the confirmation: the subject on
// the first line, then the DNS names and IP addresses the certificate is
// valid for, without duplicates and without the subject itself (compared
// ignoring case), joined by ", " on the second line. "" when the
// certificate names nothing.
func IssuedTo(c api.CertificateInfo) string {
	var lines []string
	subject := cleanText(c.Subject)
	if subject != "" {
		lines = append(lines, subject)
	}
	var names []string
	seen := map[string]bool{strings.ToLower(subject): true}
	for _, n := range append(cleanList(c.DNSNames), cleanList(c.IPAddresses)...) {
		key := strings.ToLower(n)
		if seen[key] {
			continue
		}
		seen[key] = true
		names = append(names, n)
	}
	if len(names) > 0 {
		lines = append(lines, strings.Join(names, ", "))
	}
	return strings.Join(lines, "\n")
}
